### ISSUE TITLE
GPU memory management

### ISSUE THREAD

**Issue by pranjaldhole** [yapic/yapic#50]
Currently batch size for YAPiC training cannot be dynamically adjusted based on training data.
This results in job failure if the training data batch cannot be allocated to the GPU memory even if multiple GPUs are allocated.

This issue collects and tracks ideas to handle the GPU memory management strategies.

**Comment by cmohl2013** [comment]
Currently, batch size is set automatically to number of classes, due to the following reason: In train mode, a training sample for each label class is fetched, ensuring that all label classes are present in each training step. This is beneficial for training, but not absolutely required.
For prediction result, the batch size does not matter at all (it does matter for prediction speed, though).

My proposal:
  * Add an optional argument to yapic command line interface for batch size, the default is the number of classes
  * Adjustments in traindata class: random selection of input data has to be adjusted if batch_size is not equal to number of classes. We should find some way to promote high diversity of training data (as many different classes as possible) also for smaller batch sizes. Some simple approach: 
      * If batch size is 2, ensure that at least labels from 2 classes are present.
      * If batch size is 3, ensure that at least labels from 3 classes are present.
